rust/cbork-utils/src/deterministic_helper.rs

```diff
@@ -0,0 +1,933 @@
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.key_bytes
+            .len()
+            .cmp(&other.key_bytes.len())
+            .then_with(|| self.key_bytes.cmp(&other.key_bytes))
+    }
+}
+
+/// Decodes a CBOR map with deterministic encoding validation (RFC 8949 Section 4.2.3)
+/// Returns the raw bytes of the map if it passes all deterministic validation rules.
+///
+/// From RFC 8949 Section 4.2.3:
+/// "The keys in every map must be sorted in the following order:
+///  1. If two keys have different lengths, the shorter one sorts earlier;
+///  2. If two keys have the same length, the one with the lower value in (byte-wise)
+///     lexical order sorts earlier."
+///
+/// Additionally:
+/// - Map lengths must use minimal encoding (Section 4.2.1)
+/// - Indefinite-length maps are not allowed (Section 4.2.2)
+/// - No two keys may be equal (Section 4.2.3)
+/// - The keys themselves must be deterministically encoded
+///
+/// # Errors
+///
+/// Returns `DeterministicError` if:
+/// - Input is empty (`UnexpectedEof`)
+/// - Map uses indefinite-length encoding (`IndefiniteLength`)
+/// - Map length is not encoded minimally (`NonMinimalInt`)
+/// - Map keys are not properly sorted (`UnorderedMapKeys`)
+/// - Duplicate keys are found (`DuplicateMapKey`)
+/// - Map key or value decoding fails (`DecoderError`)
+pub fn decode_map_deterministically(d: &mut Decoder) -> Result<Vec<u8>, minicbor::decode::Error> {
+    validate_input_not_empty(d)?;
+
+    // From RFC 8949 Section 4.2.2:
+    // "Indefinite-length items must be made definite-length items."
+    // The specification explicitly prohibits indefinite-length items in
+    // deterministic encoding to ensure consistent representation.
+    match d.datatype()? {
+        minicbor::data::Type::Map => {},
+        minicbor::data::Type::MapIndef => {
+            minicbor::decode::Error::message(
+                "Indefinite-length items must be made definite-length items",
+            );
+        },
+        _ => {
+            minicbor::decode::Error::message("Expected a map");
+        },
+    }
+
+    let start_pos = d.position();
+    let map_len = d.map()?.ok_or(minicbor::decode::Error::end_of_input())?;
+
+    check_map_minimal_length(d, start_pos, map_len)?;
+
+    // Store the starting position of the entire map
+    let map_start = start_pos;
+
+    // Decode entries to validate them
+    let entries = decode_map_entries(d, map_len)?;
+    validate_map_ordering(&entries)?;
+
+    // Get the ending position after validation
+    let map_end = d.position();
+
+    // Return the raw bytes of the entire validated map
+    get_map_bytes(d, map_start, map_end)
+}
+
+/// Extracts the raw bytes of a CBOR map from a decoder based on specified positions.
+/// This function retrieves the raw byte representation of a CBOR map between the given
+/// start and end positions from the decoder's underlying buffer.
+fn get_map_bytes(
```

Comment:
I think this function is generic in that sense that it just extracts bytes from the decoder and probably could be used in any scenario
```suggestion
fn get_bytes(
```